docs/advanced-recipes/atom-creators.mdx

```diff
@@ -355,4 +355,105 @@ const ListOfThings = () => {
 }
 ```
 
-<CodeSandbox id="ugkzm0" />
+## atomWithDebounce
+
+> `atomWithDebounce` helps with creating an atom where state set should be debounced.
+
+This util is useful for text search inputs, where you would like to call **functions in derived atoms only once** after waiting for a duration, instead of firing an action on every keystroke.
+
+```tsx
+import { atom, Getter, PrimitiveAtom, Setter } from 'jotai'
+import { SetStateAction } from 'react'
+
+export default function atomWithDebounce<T>(
+  initialValue: T,
+  delayMilliseconds = 500,
+  shouldDebounceOnReset = false
+) {
+  const prevTimeoutAtom = atom<ReturnType<typeof setTimeout> | null>(null)
+
+  // DO NOT EXPORT currentValueAtom as using this atom to set state can cause
+  // inconsistent state between currentValueAtom and debouncedValueAtom
+  const _currentValueAtom = atom(initialValue)
+  const isDebouncingAtom = atom(false)
+
+  const clearTimeoutId = (
+    get: Getter,
+    set: Setter,
+    prevTimeoutAtom: PrimitiveAtom<ReturnType<typeof setTimeout> | null>
+  ) => {
+    const prevTimeoutId = get(prevTimeoutAtom)
+    if (prevTimeoutId) {
+      clearTimeout(prevTimeoutId)
+      set(prevTimeoutAtom, null)
+    }
+  }
```

Comment:
This is just fine, but we prefer write-only atom pattern.

```js
const clearTimeoutAtom = atom(null, (get, set) => {
  const prevTimeout = get(prevTimeoutAtom)
  if (prevTimeout) {
    clearTimeout(prevTimeout)
    set(prevTimeoutAtom, null)
  }
})
```

In JS, `clearTimeout(null)` is no-op. So, we could simply do:
```ts
clearTimeout(get(prevTimeoutAtom) as ...)
```
But, in this recipe, the write-only atom above would be good.

Reply:
Ah, my purpose of making this function was so that it could be used in the write-only atom `clearTimeoutAtom` in line `432`, and also in the setter of `debouncedValueAtom`. Also to improve readability in the `debouncedValueAtom` setter. It was not meant to be exported.

Reply:
Amended my typings to undefined, and removed this utility function so that the signature could be cleaner. Using `undefined` instead of `null` avoids the need for any typecasting or conditional checks to make the ts linter happy.

v

Reply:
I didn't know `clearTimeout(undefined)` isn't complained by tsc.
https://tsplay.dev/wQ8rjW